File: 개인공부/hash/Camouflage.java
/*https://programmers.co.kr/learn/courses/30/lessons/42578?language=java*/
package pregrammers;

public class Camouflage {
    public static void main(String[] args) {
        String arr[][] = {{"yellow_hat", "headgear"}, {"blue_sunglasses", "eyewear"}, {"green_turban", "headgear"}};
        System.out.println(new Camouflage().solution(arr));
    }
    public int solution(final String[][] clothes) {
        int arr[] = new int[180];  //용량 30%이상 차면 해쉬충돌 확률 70%니까 3배 해줌
        int answer = 0;

        for(int i=0;i<clothes.length;i++){
            arr[makeHash(clothes[i][1],180)]++;
        }
        return NumOfCases(arr);
    }
    public int makeHash(final String source,final int size){    //해쉬만드는 중
        int ret = 0;
        int digit = 1;
        for(int i=0;i<source.length();i++){
            ret = ret + source.charAt(i) * digit + source.charAt(i);
            digit = digit * 2;
        }
        return ret%size;
    }
    public int NumOfCases(int arr[]){
        int ret = 1;
        for(int i=0;i<arr.length;i++){
            if(arr[i]>0){
                ret = ret * (arr[i]+1); //모든옷중 하나 고르는 개수 + 안고르는 경우의 수
            }
        }
        return ret-1;
    }
}

File: 알고리즘 스터디/개인공부/hash/Camouflage.py
#https://programmers.co.kr/learn/courses/30/lessons/42578

def solution(clothes):
    closet = dict()
    answer = 1
    for item, category in clothes:  # 모든 옷에 대해서 탐색
        if category in closet:  # 이전에 카운트 한 카테고리인 경우
            closet[category] += 1
        else:  # 이전에 카운트하지 않은 카테고리인 경우
            closet[category] = 1

    for key in closet:  # 모든 카테고리에 대해서 탐색
        answer *= closet[key] + 1  # 카테고리 별 옷의 개수 + 옷을 안입는 경우

    return answer - 1  # 옷을 하나도 안입는 경우를 빼줌

if __name__ == '__main__':
    solution([["yellow_hat", "headgear"], ["blue_sunglasses", "eyewear"], ["green_turban", "headgear"]])
    solution([["crow_mask", "face"], ["blue_sunglasses", "face"], ["smoky_makeup", "face"]])